src/evict/evict_page.c

```diff
@@ -663,34 +663,69 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
         if (is_eviction_thread) {
-            /*
-             * Eviction threads do not need to pin anything in the cache. We have a exclusive lock
-             * for the page being evicted so we are sure that the page will always be there while it
-             * is being processed. Therefore, we use snapshot API that doesn't publish shared IDs to
-             * the outside world.
-             */
-            __wt_txn_bump_snapshot(session);
-            snapshot_acquired = true;
-
-            /*
-             * Make sure once more that there is no checkpoint running. A new checkpoint might have
-             * started between previous check and acquiring snapshot. If there is a checkpoint
-             * running, release the snapshot and fallback to global visibility checks.
-             */
-            if (conn->txn_global.checkpoint_running) {
-                __wt_txn_release_snapshot(session);
-                snapshot_acquired = false;
+            /* Eviction thread doing eviction. */
+            if (!conn->txn_global.checkpoint_running) {
+                /*
+                 * Eviction threads do not need to pin anything in the cache. We have an exclusive
+                 * lock for the page being evicted so we are sure that the page will always be there
+                 * while it is being processed. Therefore, we use snapshot API that doesn't publish
+                 * shared IDs to the outside world.
+                 */
+                __wt_txn_bump_snapshot(session);
+                snapshot_acquired = true;
+
+                /*
+                 * Make sure once more that there is no checkpoint running. A new checkpoint might
+                 * have started between previous check and acquiring snapshot. If there is a
+                 * checkpoint running, release the snapshot and fallback to global visibility
+                 * checks.
+                 *
+                 * There should be a read barrier here otherwise the second read may be optimized
+                 * away by the compiler. However, we should have already met a barrier when we bump
+                 * the snapshot so the barrier is omitted here.
+                 */
+                if (conn->txn_global.checkpoint_running) {
+                    __wt_txn_release_snapshot(session);
+                    snapshot_acquired = false;
+                    LF_SET(WT_REC_VISIBLE_ALL | WT_REC_CHECKPOINT_RUNNING);
+                }
+            } else {
+                if (!WT_SESSION_BTREE_SYNC(session))
+                    LF_SET(WT_REC_VISIBLE_ALL);
+                LF_SET(WT_REC_CHECKPOINT_RUNNING);
+            }
+        } else if (use_snapshot_for_app_thread) {
+            /* Application thread that has a snapshot doing eviction. */
+            if (!conn->txn_global.checkpoint_running)
+                /* Use application snapshot for eviction only when checkpoint is not running. */
+                LF_SET(WT_REC_APP_EVICTION_SNAPSHOT);
```

Comment:
@quchenhao - there is still a chance that checkpoint will start running between checking the condition and setting the flag `WT_REC_APP_EVICTION_SNAPSHOT`. I wonder if we are okay with that scenario, and if yes, we need to update the comment saying so.

Reply:
I think that is OK, as checkpoint will have a newer snapshot than eviction. I'll add the comment.